### PR TITLE
"Open pull requests" page

### PR DESCRIPTION
**Pull request by petracihalova**

### CODE REVIEW

File: github_api.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    # Get list of GitHub repos from Overview page
+    services_links = routes.overview_page.get_services_links()
+    github_urls = get_github_urls(services_links)
+
+    pull_requests = {}
+    gh_token = config.GITHUB_TOKEN
+
+    # Download open pull requests
+    for url in github_urls:
+        match = re.search(pattern, url)
```
Comment (frenzymadness) [inline github_api.py:35]: If you redesign `get_github_urls` to something like `get_github_projects` and that would return just owners and projects, you would not need to use the pattern again.

File: github_api.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    return sorted(list(github_urls), key=lambda x: x.split("/")[-1])
+
+
+def get_open_pull_request():
+    # Get list of GitHub repos from Overview page
+    services_links = routes.overview_page.get_services_links()
+    github_urls = get_github_urls(services_links)
+
+    pull_requests = {}
+    gh_token = config.GITHUB_TOKEN
```
Comment (frenzymadness) [inline github_api.py:31]: What will happen if the token is not defined and therefore `None` here?

File: github_api.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+                                "html_url": pr["html_url"],
+                            }
+                        )
+
+                    pull_requests[repo_name] = open_pr_list
+
+            elif response.status_code == 401:
+                abort(401, "401 Unauthorized - check the GitHub token.")
+
+            response.raise_for_status()
```
Comment (frenzymadness) [inline github_api.py:74]: Instead `raise_for_status` and `except` you might use just `else` and `abort` in it so it won't require try and another level of indentation.

Comment (frenzymadness) [reply github_api.py:74]: But that would not catch issues with requests like timeouts or refused connections.

File: routes/open_pr_page.py [hunk 4]
```diff
@@ -0,0 +1,14 @@
+import json
+
+import config
+import github_api
+
+
+def get_github_open_pr(reload_data):
+    if not config.GITHUB_PR_LIST.is_file() or reload_data:
+        github_api.get_open_pull_request()
```
Comment (frenzymadness) [inline routes/open_pr_page.py:9]: The call to `github_api.get_open_pull_request()` can also return the PRs so you don't always need to open the cache file.

File: templates/open_pr.html [hunk 5]
```diff
@@ -2,9 +2,60 @@
...
+                <th>Draft / WIP</th>
+                <th>Author</th>
+                <th>Created at</th>
+            </tr>
+        </thead>
+        <tbody>
+            {% for repo, prs in github_open_pr.items() %}
+            {% for pr in prs %}
+            <tr>
+                <td>{{ repo}}</td>
```
Comment (frenzymadness) [inline templates/open_pr.html:27]:
```suggestion
                <td>{{ repo }}</td>
```

File: templates/open_pr.html [hunk 6]
```diff
@@ -2,9 +2,60 @@
...
+                <td>
+                    <a href="{{ pr.html_url }}" target="_blank">PR#{{ pr.number }}</a> {{ pr.title }}
+                </td>
+                <td>
+                    {% if pr.draft %}
+                    yes
```
Comment (frenzymadness) [inline templates/open_pr.html:33]: Two identical branches, can you use `or`.

File: templates/open_pr.html [hunk 7]
```diff
@@ -2,9 +2,60 @@
...
+            {% endfor %}
+            {% endfor %}
+        </tbody>
+
+    </table>
 </div>
 
+<!-- JavaScript -->
+{% block js %}
+<script src="{{ url_for('static', filename='js/open_pr.js') }}"></script>
```
Comment (frenzymadness) [inline templates/open_pr.html:50]: You might consider to include the file directly here. See https://jinja.palletsprojects.com/en/3.0.x/templates/#include

File: templates/layout.html [hunk 8]
```diff
@@ -30,12 +31,17 @@
 
         </div>
     </div>
-    <!-- Bootstrap core JS-->
+    <!-- Bootstrap, Datatables, JQuery scripts -->
     <script src="https://cdn.jsdelivr.net/npm/bootstrap@5.3.2/dist/js/bootstrap.bundle.min.js"
         integrity="sha384-C6RzsynM9kWDrMNeT87bh95OGNyZPhcTNXj1NW7RuBCsyN/o0jlpcV8Qyq46cDfL"
         crossorigin="anonymous"></script>
+    <script src="https://code.jquery.com/jquery-3.7.0.js"></script>
```
[the remark this answers is not in the document]
Comment (petracihalova) [reply templates/layout.html]: updated

File: github_api.py [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+
+def get_github_urls(links):
+    github_urls = set()
+    for category in links["categories"]:
+        for repo in category["category_repos"]:
+            for link in repo["links"]:
+                if re.match(pattern, link["link_value"]):
+                    github_urls.add(link["link_value"])
+
+    return sorted(list(github_urls), key=lambda x: x.split("/")[-1])
```
[the remark this answers is not in the document]
Comment (petracihalova) [reply github_api.py]: I didnt know, thank you ... and updated

File: templates/open_pr.html (hunk 5 above)
Comment (petracihalova) [reply templates/open_pr.html]: fixed

File: templates/open_pr.html (hunk 6 above)
Comment (petracihalova) [reply templates/open_pr.html]: fixed

File: github_api.py (hunk 1 above)
Comment (petracihalova) [reply github_api.py]: thank you, that is great idea ... updated

File: github_api.py (hunk 2 above)
Comment (petracihalova) [reply github_api.py:31]: then 401 is raised (L67)

File: github_api.py [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+                                "created_at": pr["created_at"],
+                                "user_login": pr["user"]["login"],
+                                "html_url": pr["html_url"],
+                            }
+                        )
+
+                    pull_requests[repo_name] = open_pr_list
+
+            elif response.status_code == 401:
+                abort(401, f"401 Unauthorized - check the GitHub token.")
```
Comment (frenzymadness) [inline github_api.py:67]:
```suggestion
                abort(401, "401 Unauthorized - check the GitHub token.")
```

File: app.py [hunk 11]
```diff
@@ -17,9 +27,12 @@ def deployments():
     return render_template("deployments.html")
 
 
-@app.route("/open-pr")
+@app.route("/open_pr")
 def open_pr():
-    return render_template("open_pr.html")
+    reload_data = True if request.args.get("reload_data") == "true" else False
```
Comment (frenzymadness) [inline app.py:32]: If you use 1/0 as true/false for the reload_data argument in the URL, you can get rid of this logic and use `bool(request.args.get("reload_data"))`.

Comment (petracihalova) [reply app.py]: done

File: templates/open_pr.html (hunk 7 above)
Comment (petracihalova) [reply templates/open_pr.html]: solved

File: routes/open_pr_page.py (hunk 4 above)
Comment (petracihalova) [reply routes/open_pr_page.py]: done

File: templates/open_pr.html [hunk 12]
```diff
@@ -2,9 +2,58 @@
...
+            {% endfor %}
+            {% endfor %}
+        </tbody>
+
+    </table>
 </div>
 
+<!-- JavaScript -->
+{% block js %}
+<script src="../static/js/open_pr.js"></script>
```
Comment (frenzymadness) [inline templates/open_pr.html:48]: `url_for` was better than hardcoded address and what I was proposing was using [{% include %}](https://jinja.palletsprojects.com/en/3.0.x/templates/#include).

**Review by frenzymadness — COMMENTED** [review]
One little misunderstanding, the rest look good to me.

File: templates/open_pr.html (hunk 12 above)
Comment (petracihalova) [reply templates/open_pr.html]: I think the `include` tag is used only for html templates https://jinja.palletsprojects.com/en/3.0.x/templates/?highlight=include#include

Comment (frenzymadness) [reply templates/open_pr.html]: I believe you can use Jinja2 for whatever text files you want. Using include here is not a hard requirement so feel free to just reverse the change back to the url_for and forget about the idea.

Comment (petracihalova) [reply templates/open_pr.html]: done

**Pull request merged**